service/controller/app/v1/resource_set.go

```diff
@@ -159,19 +175,34 @@ func NewResourceSet(config ResourceSetConfig) (*controller.ResourceSet, error) {
 			return nil, microerror.Mask(err)
 		}
 
-		g8sClient, err := kubeConfig.NewG8sClientForApp(ctx, cr)
-		if err != nil {
-			return nil, microerror.Mask(err)
-		}
+		secretName := key.SecretName(cr)
 
-		k8sClient, err := kubeConfig.NewK8sClientForApp(ctx, cr)
-		if err != nil {
-			return nil, microerror.Mask(err)
+		var restConfig *rest.Config
+		var g8sClient versioned.Interface
+		var k8sClient kubernetes.Interface
+
+		if secretName != "" {
```

Comment:
It is a bit odd that the secret name decides when to initialize the clients but the interface of initializing the clients does not seem to be concerned with the secret name. That smells a bit.

Reply:
As explained in here, https://github.com/giantswarm/app-operator/pull/33#discussion_r254638721, `secret` information is necessary to access to tenant cluster.

Reply:
I understood this, but the decision making the the interface diverge so there are different parts doing the same thing, which is just not obvious in the first place.